src/test/regress/sql/merge.sql

```diff
@@ -824,10 +861,716 @@ RESET client_min_messages;
+
+MERGE INTO citus_pa_target t
+  USING citus_pa_source s
+  ON t.tid = s.sid
+  WHEN MATCHED THEN
+    UPDATE SET balance = balance + delta, val = val || ' updated by merge'
+  WHEN NOT MATCHED THEN
+    INSERT VALUES (sid, delta, 'inserted by merge');
+
+SELECT pa_compare_tables();
+ROLLBACK;
+
+-- same with a constant qual
+BEGIN;
+MERGE INTO pg_pa_target t
+  USING pg_pa_source s
+  ON t.tid = s.sid AND tid = 1
+  WHEN MATCHED THEN
+    UPDATE SET balance = balance + delta, val = val || ' updated by merge'
+  WHEN NOT MATCHED THEN
+    INSERT VALUES (sid, delta, 'inserted by merge');
+
+MERGE INTO citus_pa_target t
+  USING citus_pa_source s
+  ON t.tid = s.sid AND tid = 1
+  WHEN MATCHED THEN
+    UPDATE SET balance = balance + delta, val = val || ' updated by merge'
+  WHEN NOT MATCHED THEN
+    INSERT VALUES (sid, delta, 'inserted by merge');
+
+SELECT pa_compare_tables();
+ROLLBACK;
+
+CREATE TABLE source_json( id   integer, z int, d jsonb);
+CREATE TABLE target_json( id   integer, z int, d jsonb);
+
+INSERT INTO source_json SELECT i,i FROM generate_series(0,100)i;
+
+SELECT create_distributed_table('target_json','id'), create_distributed_table('source_json', 'id');
+
+-- single shard query given source_json is filtered and Postgres is smart to pushdown
+-- filter to the target_json as well
+EXPLAIN MERGE INTO target_json sda
+USING (SELECT * FROM source_json WHERE id = 1) sdn
+ON sda.id = sdn.id
+WHEN NOT matched THEN
+	INSERT (id, z) VALUES (sdn.id, 5);
+
+-- zero shard query as filters do not match
+EXPLAIN MERGE INTO target_json sda
+USING (SELECT * FROM source_json WHERE id = 1) sdn
+ON sda.id = sdn.id AND sda.id = 2
+WHEN NOT matched THEN
+	INSERT (id, z) VALUES (sdn.id, 5);
+
+-- join for source_json is happening at at different place
+EXPLAIN MERGE INTO target_json sda
+USING source_json s1 LEFT JOIN (SELECT * FROM source_json) s2 USING(z)
+ON sda.id = s1.id AND s1.id = s2.id
+WHEN NOT matched THEN
+	INSERT (id, z) VALUES (s2.id, 5);
+
+-- update JSON column
+EXPLAIN MERGE INTO target_json sda
+USING source_json sdn
+ON sda.id = sdn.id
+WHEN matched THEN
+	UPDATE SET d = '{"a" : 5}';
+
+CREATE FUNCTION immutable_hash(int) RETURNS int
+AS 'SELECT hashtext( ($1 + $1)::text);'
+LANGUAGE SQL
+IMMUTABLE
+RETURNS NULL ON NULL INPUT;
+
+MERGE INTO target_json sda
+USING source_json sdn
+ON sda.id = sdn.id
+WHEN matched THEN
+	UPDATE SET z = immutable_hash(sdn.z);
+
+-- Test bigserial
+CREATE TABLE source_serial (id integer, z int, d bigserial);
+CREATE TABLE target_serial (id integer, z int, d bigserial);
+INSERT INTO source_serial SELECT i,i FROM generate_series(0,100)i;
+SELECT create_distributed_table('source_serial', 'id'), create_distributed_table('target_serial', 'id');
+
+MERGE INTO target_serial sda
+USING source_serial sdn
+ON sda.id = sdn.id
+WHEN NOT matched THEN
+       INSERT (id, z) VALUES (id, z);
+
+SELECT count(*) from source_serial;
+SELECT count(*) from target_serial;
```

Comment:
I think we have some problems with serials. In the local case, we get distinct `d` values on the target, whereas on the distributed case, we get a single value:

```SQL
CREATE TABLE source_serial (id integer, z int, d bigserial);
CREATE TABLE target_serial (id integer, z int, d bigserial);
INSERT INTO source_serial SELECT i,i FROM generate_series(0,100)i;

MERGE INTO target_serial sda
USING source_serial sdn
ON sda.id = sdn.id
WHEN NOT matched THEN
       INSERT (id, z) VALUES (id, z);

SELECT count(distinct d) from target_serial;
┌───────┐
│ count │
├───────┤
│   101 │
└───────┘
(1 row)


TRUNCATE target_serial, source_serial;

SELECT create_distributed_table('source_serial', 'id'), create_distributed_table('target_serial', 'id');

INSERT INTO source_serial SELECT i,i FROM generate_series(0,100)i;

MERGE INTO target_serial sda
USING source_serial sdn
ON sda.id = sdn.id
WHEN NOT matched THEN
       INSERT (id, z) VALUES (id, z);


SELECT count(distinct d) from target_serial;
┌───────┐
│ count │
├───────┤
│     1 │
└───────┘
(1 row)
```

I currently don't have time to investigate, but let me know if this part of the code is not familiar for you to debug on what is going on.

Reply:
We also shouldn't be allowing:
```SQL
 MERGE INTO target_serial sda                                                                                         USING source_serial sdn                                                                                                                                        ON sda.id = sdn.id                                                                                                                                             WHEN NOT matched THEN                                                                                                                                                 INSERT (id, z) VALUES (id, nextval('numbers'));
```

I wonder why the checks in `MergeActionListSupported` not triggered?

Reply:
The crux of the problem is, we are issuing MERGE with the value `102` (the next serial value)

`NOTICE:  issuing MERGE INTO test.target_serial_102330 first USING test.source_serial_102326 sdn ON (first.id OPERATOR(pg_catalog.=) sdn.id) WHEN NOT MATCHED THEN INSERT (id, z, d) VALUES (sdn.id, sdn.z, '102'::bigint)`

Very interesting case, during the planning phase we do have the `TargetEntry` of `INSERT `as the `T_FuncExpr`, but during the run time, it's `T_Const` with the value `102`. The reason why we deparse at runtime (and not planning time) is because of the flag `RequiresCoordinatorEvaluation` which is true if the query contains any stable or volatile function. The `bigserial` function is causing this flag to true. 

Let me debug this further to see how we can solve this problem.

Reply:
Under debugger, I've forcefully set `RequiresCoordinatorEvaluation` to false and we get correct results

 ```
count
-------
   101
(1 row)
```

`NOTICE:  issuing MERGE INTO test.target_serial_102341 first USING test.source_serial_102337 sdn ON (first.id OPERATOR(pg_catalog.=) sdn.id) WHEN NOT MATCHED THEN INSERT (id, z, d) VALUES (sdn.id, sdn.z, nextval('test.target_serial_d_seq'::regclass))
`

Reply:
OK, this is the similar problem I faced in `TargetlistAndFunctionsSupported` and I have to fanout a new routine to address such issues. `RequiresCoordinatorEvaluation` relies on the logic that any function present in the entire query tree might need a coordinator evalution, but this logic doesn't fully work for MERGE which has multiple query-statements in merge-actions. A function present in `INSERT/UPDATE/DELETE` not necessarily mean the top-level query needs coordinator evaluation. I fixed it now.

Reply:
BTW, this is a very good catch, very sneaky one to detect.

Reply:
Regarding the other comment

>>I wonder why the checks in MergeActionListSupported not triggered?

```
MERGE INTO target_serial sda
USING source_serial sdn
ON sda.id = sdn.id
WHEN NOT matched THEN
INSERT (id, z) VALUES (id, nextval('numbers'));

```

We used to block it in `MergeActionListSupported` but unfortunately that became too restrictive, a simple `INSERT`, such as,

`INSERT VALUES (slogts::timestamp, sid, delta, 'inserted by merge');
`

 is also getting blocked. I looked around the code, we are doing such function checks only for `UPDATE/DELETE`, also if we restrict volatile/stable functions in `INSERT`, the serial column `INSERTS `too will block.

Reply:
Here is an example of it, vanilla INSERT is not blocked

```
postgres=# INSERT into target_serial(id, z) VALUES (300, nextval('numbers_j_seq'));
INSERT into target_serial(id, z) VALUES (300, nextval('numbers_j_seq'));
NOTICE:  issuing INSERT INTO test.target_serial_103489 (id, z, d) VALUES (300, 2, '3'::bigint)
DETAIL:  on server citus@127.0.0.1:11000 connectionId: 11
INSERT 0 1

```

Reply:
there are a lot of parallels with INSERT..SELECT, and this is also a limitation of INSERT..SELECT pushdown

where we'd like to be is that when something like nextval is used we fall back to a MERGE implementation that inserts via the coordinator.

that said, we supported INSERT..SELECT pushdown before we supported INSERT..SELECT via coordinator, so for now it's probably ok to just error when there are volatile functions. Ideally, write the check in a way where we can easily switch to another implementation later (e.g. using DeferredError?)

Reply:
Yes, if we pushdown nextval(..) is getting evaluated at the shard(I've discussed this in our engine team chat). I'm blocking volatile functions in the actions part of the Merge (we already have such block for the Merge query itself) too.

Reply:
For now, we blocked all non-immutable functions, we will revisit this in the coming PRs